src/spinnaker.cpp

```diff
@@ -0,0 +1,1033 @@
+{
+    try {
+        CHECK(self_);
+        ((struct SpinnakerCamera*)self_)->get_shape(shape);
+        return Device_Ok;
+    } catch (const std::exception& exc) {
+        LOGE("Exception: %s\n", exc.what());
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+    return Device_Err;
+}
+
+enum DeviceStatusCode
+spinnakercam_stop(struct Camera* self_)
+{
+    try {
+        CHECK(self_);
+        ((struct SpinnakerCamera*)self_)->stop();
+        return Device_Ok;
+    } catch (const std::exception& exc) {
+        LOGE("Exception: %s\n", exc.what());
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+    return Device_Err;
+}
+
+enum DeviceStatusCode
+spinnakercam_start(struct Camera* self_)
+{
+    try {
+        CHECK(self_);
+        ((struct SpinnakerCamera*)self_)->start();
+        return Device_Ok;
+    } catch (const std::exception& exc) {
+        LOGE("Exception: %s\n", exc.what());
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+    return Device_Err;
+}
+
+enum DeviceStatusCode
+spinnakercam_execute_trigger(struct Camera* self_)
+{
+    try {
+        CHECK(self_);
+        ((struct SpinnakerCamera*)self_)->execute_trigger();
+        return Device_Ok;
+    } catch (const std::exception& exc) {
+        LOGE("Exception: %s\n", exc.what());
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+    return Device_Err;
+}
+
+enum DeviceStatusCode
+spinnakercam_get_frame(struct Camera* self_,
+                       void* im,
+                       size_t* nbytes,
+                       struct ImageInfo* info)
+{
+    try {
+        CHECK(self_);
+        ((struct SpinnakerCamera*)self_)->get_frame(im, nbytes, info);
+        return Device_Ok;
+    } catch (const std::exception& exc) {
+        LOGE("Exception: %s\n", exc.what());
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+    return Device_Err;
+}
+
+//
+// Camera implementation
+//
+
+SpinnakerCamera::SpinnakerCamera(Spinnaker::CameraPtr camera)
+  : Camera{ .set = ::spinnakercam_set,
+            .get = ::spinnakercam_get,
+            .get_meta = ::spinnakercam_get_meta,
+            .get_shape = ::spinnakercam_get_shape,
+            .start = ::spinnakercam_start,
+            .stop = ::spinnakercam_stop,
+            .execute_trigger = ::spinnakercam_execute_trigger,
+            .get_frame = ::spinnakercam_get_frame,
+  }
+  , camera_(camera)
+  , last_known_settings_{}
+  , frame_id_(0)
+  , started_(false)
+{
+    // Sometimes the camera is still initialized from a previous run.
+    // Ideally, we would error here, but that can make the camera indefinitely
+    // unusable in Acquire, so log and reset instead.
+    if (camera->IsInitialized()) {
+        LOGE("Camera was already initialized. De-initializing it before "
+             "re-initializing.");
+        camera_->DeInit();
+    }
+    camera->Init();
```

Comment:
Sometimes spinnaker can throw here, complaining that the device does not have enough memory for buffer allocation. Changing the buffer handling mode to newest only seemed to help, but did not entirely fix things.

```cpp
    Spinnaker::GenApi::INodeMap& stream_node_map = camera->GetTLStreamNodeMap();
    Spinnaker::GenApi::CEnumerationPtr stream_handling_node = stream_node_map.GetNode("StreamBufferHandlingMode");
    *stream_handling_node = "NewestOnly";
```

Reply:
It might be worth being opinionated about the stream handling so we get consistent behavior in the field.